plugins/server_logs/lib/samson_server_logs/kafka_listener.rb

```diff
@@ -0,0 +1,120 @@
+require 'poseidon'
+
+module SamsonServerLogs
+  class KafkaListener
+    TOTAL_TIMEOUT = (Rails.env.test? ? 1.second : 30.minutes)
+    AFTER_DEPLOY_TIMEOUT = (Rails.env.test? ? 1.second : 10.minutes)
+    KAFKA_HOST = "kafka1.pod100" # TODO dynamic
+    KAFKA_PORT = 9092
+    KAFKA_TOPIC = "samson_deploy"
+
+    def initialize(deploy_id, output)
+      @deploy_id = deploy_id
+      @output = output
+      @consumer = Poseidon::PartitionConsumer.new(
+        "samson_deploy_listener",
+        KAFKA_HOST, KAFKA_PORT,
+        KAFKA_TOPIC, 0, :latest_offset
+      )
+    end
+
+    def listen
+      config = {
+        timeout: TOTAL_TIMEOUT.from_now,
+        servers: nil,
+        print: false
+      }
+
+      kafka = listen_to_kafka(config)
+      server_listener = listen_for_server_list(config)
+
+      deploy_success = yield
+
+      # no servers found during deploy ?
+      unless config.fetch(:servers)
+        kafka.kill
+        server_listener.kill
+        write "did not find any servers in deploy log"
+        return false
+      end
+
+      # wait for this time after the deploy is done
+      config[:timeout] = AFTER_DEPLOY_TIMEOUT.from_now
+      config[:print] = :all
+
+      return false unless kafka.value
+
+      deploy_success
+    end
+
+    private
+
+    # listen for the line that tells us which servers we actually deploy to
+    def listen_for_server_list(config)
+      Thread.new do
+        @output.each do |_, message|
+          if servers = message[/uploading deploy logs: \S+ servers: (\[.*?\])/i, 1]
+            config[:servers] = JSON.load(servers)
+            break
+          end
+        end
+      end
+    end
+
+    def listen_to_kafka(config)
+      Thread.new do
+        messages = []
+        done_servers = []
+        loop do
+          received = listen_for_messages(messages)
+
+          print = decide_what_to_print(config, received, messages)
+          print.each { |m| write m }
+
+          done_servers.concat(print.map { |m| m["host"] if m["exit"] == 0 }) # TODO handle other exit codes ...
+
+          # stop when we timeout or all servers are done
+          if Time.now > config.fetch(:timeout)
+            write "listener timed out"
+            break
+          elsif waiting_for_servers = config.fetch(:servers)
+            done_servers.each { |s| waiting_for_servers.delete(s) }
+            done_servers.clear
+
+            if waiting_for_servers.none?
+              write "received messages from all hosts"
+              break
+            end
+          end
+        end
+
+        (config.fetch(:servers) || []).empty? # did we complete ?
```

Comment:
it will be false if there are servers left that did not report done, changed that to `@servers && @servers.empty?` so that it's false when no servers where found too